ggml-rpc: chunk send()/recv() to avoid EINVAL for very large tensors over RPC (macOS & others)

Tak-RS opened the pull request:
Fixes #15055

This PR prevents send()/recv() from being called with extremely large buffers during RPC tensor transfer by chunking I/O into 1 GiB pieces. On macOS this avoids intermittent EINVAL errors that previously caused the client to abort when offloading very large models via RPC.

What’s the symptom?
Loading very large GGUFs via RPC would fail with:

client: send: Invalid argument

server: recv: Invalid argument

Repro seen with DeepSeek-R1-0528-* and Qwen3-480B-* at large quants.

Single-node load (no RPC) worked fine.

Root cause (observed)
OS-level limits on single send()/recv() buffer sizes; very large tensors were transmitted in one shot. Splitting into smaller chunks resolves the issue.

Changes
Add RPC_IO_CHUNK = 1 GiB.

Update send_data() and recv_data() to loop with chunked I/O.

Keep existing error logging; behavior is otherwise unchanged.

Why 1 GiB?
Empirically under the limits that triggered EINVAL on macOS.

Large enough to keep throughput good; easy to tune later if needed.

Testing
macOS (Metal): Previously failing large-model RPC offload now completes. Inference runs.

macOS (non-Metal): Build + basic RPC transfer OK.

Linux/Ubuntu: Not tested yet. Relying on CI and maintainer validation. (Happy to test on request; I can also try Docker later.)

Known quirk (non-blocking)
I still see an occasional non-fatal recv: Invalid argument before the big tensor transfer starts, but the run proceeds and finishes. I suspect a minor size-field mismatch during early handshake. If useful, I can follow up with a tiny patch that always serializes message lengths as uint64_t on the wire.

Performance / compatibility
No API changes.

Chunking is per-call looped send/recv; negligible overhead in my tests vs. “one big send”.

Should be safe across platforms.

Thanks!

---

Update: Verified cross-OS direction as well.

**Additional testing**
- Client: Ubuntu 22.04 (glibc), clang/gcc build, commit 0e7aa4e
- RPC server: macOS (Apple M3 Ultra, 512 GB RAM, Metal enabled)
- llama.cpp build: Release
- Model: DeepSeek-R1-0528-Q4_K_M (GGUF format, very large tensor size)
- Command (client):
  ./build/bin/llama-server \
    -m /path/to/DeepSeek-R1-0528-Q4_K_M.gguf \
    --rpc <mac-ip>:50052 -c 3000
- Command (server):
  ./build/bin/rpc-server -p 50052 --host <mac-ip>

**Result**
- Large tensor offload succeeds end-to-end. Inference runs normally.
- No client/server aborts observed.
- Occasionally still see a *non-fatal* `recv: Invalid argument` before the first large transfer; run proceeds normally.

**Notes**
- The chunked I/O change fixes the main crash.  
- If you prefer, I can follow up with a tiny patch to always serialize message length fields as `uint64_t` to silence the early handshake warning.

Commit:
ggml-rpc: chunk send()/recv() to avoid EINVAL for very large tensors over RPC (macOS & others). Fixes #15055

rgerganov reviewed (changes requested):
Thank you for the bug report and the patch

Tak-RS committed:
ggml-rpc: rename RPC_IO_CHUNK->MAX_CHUNK_SIZE, use std::min() for cap, switch to GGML_LOG_ERROR, handle 0-length send/recv

Tak-RS commented:
Thank you for the review and suggestions!

Applied the requested changes:
- Renamed `RPC_IO_CHUNK` → `MAX_CHUNK_SIZE`
- Switched error logging from `perror`/`fprintf` to `GGML_LOG_ERROR`
- Corrected chunk size calculation to use `std::min(size - bytes_sent, MAX_CHUNK_SIZE)` (cap instead of max)
- Same fix applied in `recv_data()`
- Added a check to treat 0-length send/recv as an error

Please let me know if you’d like further changes.

rgerganov commented on ggml/src/ggml-rpc/ggml-rpc.cpp:
```diff
@@ -323,27 +326,45 @@ static std::shared_ptr<socket_t> create_server_socket(const char * host, int por
+                           bytes_sent, size_to_send);
+            return false;
+        }
+        if (n == 0) {
```
why do we need this special case for `n == 0`? if zero bytes are sent, then we should retry again until we send everything or an error occurs (`n < 0`)

Tak-RS replied:
Got it — I’ll remove the special case for `n == 0` in send_data() 
and just retry in the loop as suggested.

Tak-RS committed:
rpc: drop n==0 special case in send_data(); retry in loop per review

Tak-RS committed:
rpc: remove trailing whitespace in send_data()
